From df757ecd47e17609201612150774d493f35a1039 Mon Sep 17 00:00:00 2001
From: Tomasz Figa <tfiga@chromium.org>
Date: Mon, 25 Jul 2016 20:54:32 +0900
Subject: [PATCH 2/2] HACK: egl/android: Handle HAL_PIXEL_FORMAT_YCbCr_420_888
 as YV12

We don't have proper API to query such buffers for the real internal
format in a generic, scalable way, so let's just hardcode this format to
YV12 for the time being, in the same way as we hardcoded it in drm_gralloc
for all platforms using Mesa.

TODO: Revert when we have the necessary API in place (see b/29886289).

BUG=b:29061386
TEST=CTS:MediaCodecCapabilitiesTest#testAllNonTunneledVideoCodecsSupportFlexibleYUV

Change-Id: If19199c695335d6c81fbf1839f4c8f7fc2b32cc1
Signed-off-by: Tomasz Figa <tfiga@chromium.org>
Reviewed-on: https://chromium-review.googlesource.com/362736
Reviewed-by: Owen Lin <owenlin@chromium.org>
(cherry picked from commit 1de4518c9912ea7b47538ac794320b3c689904e5)
Signed-off-by: Tomasz Figa <tfiga@chromium.org>
---
 src/egl/drivers/dri2/platform_android.c | 15 +++++++++++++++
 1 file changed, 15 insertions(+)

diff --git a/src/egl/drivers/dri2/platform_android.c b/src/egl/drivers/dri2/platform_android.c
index f0d6a9e..d5a90b0 100644
--- a/src/egl/drivers/dri2/platform_android.c
+++ b/src/egl/drivers/dri2/platform_android.c
@@ -67,6 +67,11 @@ get_format_bpp(int native)
    case HAL_PIXEL_FORMAT_RGB_565:
       bpp = 2;
       break;
+   case HAL_PIXEL_FORMAT_YCbCr_420_888:
+      /*
+       * HACK: Hardcode this to YV12 as gralloc does for platforms using Mesa.
+       * TODO: Revert this once b/29886289 is fixed.
+       */
    case HAL_PIXEL_FORMAT_YV12:
       bpp = 1;
       break;
@@ -91,6 +96,11 @@ static int get_fourcc(int native)
        * TODO: Revert this once b/29886289 is fixed.
        */
    case HAL_PIXEL_FORMAT_RGBX_8888: return __DRI_IMAGE_FOURCC_XBGR8888;
+   case HAL_PIXEL_FORMAT_YCbCr_420_888:
+      /*
+       * HACK: Hardcode this to YV12 as gralloc does for platforms using Mesa.
+       * TODO: Revert this once b/29886289 is fixed.
+       */
    case HAL_PIXEL_FORMAT_YV12:      return __DRI_IMAGE_FOURCC_YVU420;
    default:
       _eglLog(_EGL_WARNING, "unsupported native buffer format 0x%x", native);
@@ -577,6 +587,11 @@ droid_create_image_from_prime_fd(_EGLDisplay *disp, _EGLContext *ctx,
       return NULL;
 
    switch (buf->format) {
+   case HAL_PIXEL_FORMAT_YCbCr_420_888:
+      /*
+       * HACK: Hardcode this to YV12 as gralloc does for platforms using Mesa.
+       * TODO: Revert this once b/29886289 is fixed.
+       */
    case HAL_PIXEL_FORMAT_YV12:
       /* Y plane is at offset 0 and fds[0] is already set. */
       /* Cr plane is located after Y plane */
-- 
2.8.0.rc3.226.g39d4020

